### PR TITLE
add lombok model support on spring

### PR DESCRIPTION
**Pull request by dabdirb**
There are several issues raised to request generate model with lombok support, include #5447, #10841, #14156, #16250, #17606
This PR add lombok support on spring project, i plan to first implement it on spring then port to all other java projects.
When user add lombok annotation by `additionalModelTypeAnnotations`, 
```sh
java -jar openapi-generator-cli.jar generate -g spring \
  -i petstore.json \
 --additional-properties=hideGenerationTimestamp=true,additionalModelTypeAnnotations=@lombok.Data;@lombok.NoArgsConstructor;@lombok.AllArgsConstructor
```
will generate below code, clean with just field definations.
```java
/**
 * Tag
 */
@lombok.Data
@lombok.NoArgsConstructor
@lombok.AllArgsConstructor

@Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
public class Tag {

  private Long id;

  private String name;

}
```
User can combine the lombok annotations to customize the output model
```sh
java -jar openapi-generator-cli.jar generate -g spring \
  -i petstore.json \
  --additional-properties=hideGenerationTimestamp=true,additionalModelTypeAnnotations=@lombok.RequiredArgsConstructor;@lombok.NoArgsConstructor;@lombok.AllArgsConstructor
```
```java

/**
 * Tag
 */
@lombok.RequiredArgsConstructor
@lombok.NoArgsConstructor
@lombok.AllArgsConstructor

@Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
public class Tag {

  private Long id;

  private String name;

  public Tag id(Long id) {
    this.id = id;
    return this;
  }

  /**
   * Get id
   * @return id
  */
  
  @Schema(name = "id", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
  @JsonProperty("id")
  public Long getId() {
    return id;
  }

  public void setId(Long id) {
    this.id = id;
  }

  public Tag name(String name) {
    this.name = name;
    return this;
  }

  /**
   * Get name
   * @return name
  */
  
  @Schema(name = "name", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
  @JsonProperty("name")
  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Tag tag = (Tag) o;
    return Objects.equals(this.id, tag.id) &&
        Objects.equals(this.name, tag.name);
  }

  @Override
  public int hashCode() {
    return Objects.hash(id, name);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class Tag {\n");
    sb.append("    id: ").append(toIndentedString(id)).append("\n");
    sb.append("    name: ").append(toIndentedString(name)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  private String toIndentedString(Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }
}
```
### Note
This implementation will discard the `x-setter-extra-annotation` in setter function and swagger `@Schema` in geter function.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by AhmedAbdelfaheem** [comment]
When this fix will be merged ?

**Comment by wing328** [comment]
cc 
@cachescrubber (2022/02) @welshm (2022/02) @MelleD (2022/02) @atextor (2022/02) @manedev79 (2022/02) @javisst (2022/02) @borsch (2022/02) @banlevente (2022/02) @Zomzog (2022/09) @martin-mfg (2023/08)

**Review by martin-mfg — APPROVED** [review]
LGTM

Thanks a lot for this PR!

-------------

Using lombok will change the format of `toString()`s output, which some users might see as disadvantage. But since using lombok is opt-in, I think that's not a problem.
e.g. from
```js
class MyError {
    message: message
    code: 1
    innerObj: class MyErrorInnerObj {
        innerMsg: innerMsg
    }
```
to 
```js
MyError(message=message, code=1, innerObj=MyErrorInnerObj(innerMsg=innerMsg))
```

**Review by welshm — COMMENTED** [review]
Thanks the feature looks great!

Would be good to add tests for the Lombok feature.

**Comment by welshm** [comment]
Is NoArgsConstructor still required to work properly with Spring? I know it used to be - and if that's the case might be worth documenting as part of the feature that's generally suggested that particular annotation always be included

**Comment by dabdirb** [comment]
@welshm NoArgsConstructor  does not required
The toString() output format is diff, in fact legacy toString() works well, if a developer want to use @lombok.ToString, he intend to replace it.
I have added a UT case to cover the lombok annotation parsing.

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java [hunk 1]
```diff
@@ -4427,4 +4427,48 @@ public void testMultiInheritanceParentRequiredParams_issue15796() throws IOExcep
                 .hasParameter("race").toConstructor()
         ;
     }
+
+    @Test
+    public void testLombokAnnotations() throws IOException {
```
Comment (welshm) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java:4432]: Beautiful tests - thank you!

🙌

**Review by welshm — APPROVED** [review]

**Comment by wing328** [comment]
@dabdirb thanks for the PR.

What about adding a config under ./bin/configs/ to have this feature enabled and test the sample in the github workflow?
```
.github/workflows/samples-spring-jdk17.yaml
.github/workflows/samples-spring.yaml
```

for test spec, you can use https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/petstore.yaml to start with

**Comment by wing328** [comment]
lgtm. thanks for the PR.

**Comment by wing328** [comment]
@dabdirb can you please PM me via Slack when you've time next week?

https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g

Have a nice weekend.

**Comment by switang** [comment]
@dabdirb Thanks for adding this! We were able to reduce some additional config now that the `lombok.*` are available for conditions.

I include this over the `var` output in the pojo file since we like to use the Builder annotation:
`{{#defaultValue}}{{#lombok.Builder}}@lombok.Builder.Default{{/lombok.Builder}}{{/defaultValue}}`

Used to have an additional property that we set and used in the pojo, but now just reference `lombok.Builder`

**Pull request merged**